libminifi/test/unit/NetworkPrioritizerServiceTests.cpp

```diff
@@ -41,129 +47,105 @@ TEST_CASE("TestPrioritizerOneInterface", "[test1]") {
-  auto controller3 = std::make_shared<minifi::controllers::NetworkPrioritizerService>("TestService3");
-  std::shared_ptr<minifi::Configure> configuration = std::make_shared<minifi::Configure>();
-  controller->initialize();
-  controller->setProperty(minifi::controllers::NetworkPrioritizerService::VerifyInterfaces, "false");
+  auto controller1 = createNetworkPrioritizerService("TestService_eth1", clock);
+  controller1->initialize();
+  controller1->setProperty(minifi::controllers::NetworkPrioritizerService::NetworkControllers, "eth1");
+  controller1->setProperty(minifi::controllers::NetworkPrioritizerService::VerifyInterfaces, "false");
+  controller1->setProperty(minifi::controllers::NetworkPrioritizerService::MaxThroughput, "10 B");
+  controller1->onEnable();
 
-  controller3->initialize();
-  controller3->setProperty(minifi::controllers::NetworkPrioritizerService::NetworkControllers, "eth0");
-  controller3->setProperty(minifi::controllers::NetworkPrioritizerService::VerifyInterfaces, "false");
-  controller3->setProperty(minifi::controllers::NetworkPrioritizerService::MaxThroughput, "1 kB");
-  controller3->onEnable();
-
-  controller2->initialize();
-  controller2->setProperty(minifi::controllers::NetworkPrioritizerService::NetworkControllers, "eth1");
-  controller2->setProperty(minifi::controllers::NetworkPrioritizerService::VerifyInterfaces, "false");
-  controller2->setProperty(minifi::controllers::NetworkPrioritizerService::MaxThroughput, "10 B");
-  controller2->onEnable();
   std::vector<std::shared_ptr<core::controller::ControllerService> > services;
-  services.push_back(controller3);
-  services.push_back(controller2);
-  controller->setLinkedControllerServices(services);
-  controller->onEnable();
-  // can't because we've triggered the max payload
-  for (int i = 0; i < 50; i++) {
-    REQUIRE("eth0" == controller->getInterface(5).getInterface());
-    REQUIRE("eth0" == controller->getInterface(5).getInterface());
-    REQUIRE("eth0" == controller->getInterface(5).getInterface());
-    REQUIRE("eth0" == controller->getInterface(5).getInterface());
-    std::this_thread::sleep_for(std::chrono::milliseconds(10));
+  services.push_back(controller0);
+  services.push_back(controller1);
+  parent_controller->setLinkedControllerServices(services);
+  parent_controller->onEnable();
+
+  SECTION("Switch to second interface when the first is saturated") {
+    REQUIRE("eth0" == parent_controller->getInterface(5).getInterface());
+    REQUIRE("eth0" == parent_controller->getInterface(5).getInterface());
+    // triggered the max throughput on eth0, switching to eth1
+    REQUIRE("eth1" == parent_controller->getInterface(5).getInterface());
+    REQUIRE("eth1" == parent_controller->getInterface(5).getInterface());
   }
-}
 
+  SECTION("Can keep sending on eth0 if we wait between packets") {
+    for (int i = 0; i < 100; i++) {
+      REQUIRE("eth0" == parent_controller->getInterface(10).getInterface());
```

Comment:
this is a little unclear to me, so `getInterface(size)` returns a network interface that we should use to send the next `size` number of bytes? what controls how long we should wait (5ms in this case) until the same interface is available?

Reply:
> `getInterface(size)` returns a network interface that we should use to send the next size number of bytes?

Yes.

> what controls how long we should wait (5ms in this case) until the same interface is available?

It is a (version of the) [token bucket algorithm](https://en.wikipedia.org/wiki/Token_bucket): conceptually, the bucket starts out with `b` tokens and `t` tokens are added to the bucket each millisecond, and each send operation removes `s` times `size` tokens from the bucket, if there are enough tokens available, otherwise it fails (and does not remove any tokens).

[In practice, the bucket is filled in `getInterface()` according to the number of elapsed milliseconds since the last call to `getInterface()`, and the way the `b`, `t` and `s` parameters are set is a bit strange (`t` is always 2, `b` and `s` depend on `MaxThroughput`, but differently depending on whenter `MaxThroughput` is less or more than 1 kB).]

Since I don't think anybody actually uses this class at the moment, I did not clean it up; I only fixed the unit test so that it is no longer dependent on timing, and fixed a few obvious bugs in the service code.

Reply:
thank you for clarifying!